tests/apps/test_train.py

```diff
@@ -0,0 +1,213 @@
+
+    def test_A(self, adj, n_mean, params):
+        """Test that the A and _A_scaled methods correctly return the WAW matrix of the A_init
+        and A_init_scaled adjacency matrices, respectively, and that these outputs are different
+        (due to the scaling)"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        A = gbs.A(params)
+        A_scaled = gbs._A_scaled(params)
+        W = gbs.W(params)
+
+        assert np.allclose(A, W @ adj @ W)
+        assert np.allclose(A_scaled, W @ gbs.A_init_scaled @ W)
+        assert not np.allclose(A, A_scaled)
+
+    def test_generate_samples(self, adj, n_mean, monkeypatch):
+        """Test that _generate_samples correctly dispatches between torontonian and hafnian
+        sampling based upon whether threshold=True or threshold=False"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "torontonian_sample_state", lambda *args, **kwargs: 0)
+            s_threshold = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        gbs.threshold = False
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "hafnian_sample_state", lambda *args, **kwargs: 1)
+            s_pnr = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        assert s_threshold == 0
+        assert s_pnr == 1
+
+    def test_add_A_init_samples_bad_shape(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples raises a ValueError when input samples of incorrect
+        shape, i.e. of dim + 1 modes"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim + 1))
+        with pytest.raises(ValueError, match="Must input samples of shape"):
+            gbs._add_A_init_samples(s)
+
+    def test_add_A_init_samples_none_there(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples correctly adds samples"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim))
+        gbs._add_A_init_samples(s)
+        assert np.allclose(gbs._A_init_samples, s)
+
+    def test_add_A_init_samples_already_there(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples correctly adds more samples when some are already there"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim))
+        gbs._A_init_samples = s
+        gbs._add_A_init_samples(s)
+        assert gbs._A_init_samples.shape == (4, dim)
+
+    def test_mean_photons_per_mode(self, n_mean, dim):
+        """Test that mean_photons_by_mode is correct when given a simple fully connected
+        adjacency matrix and an identity W. We expect each mode to have the same mean photon number
+        and for that to add up to n_mean."""
+        adj = np.ones((dim, dim))
+        params = np.zeros(dim)
+        gbs = train.VGBS(adj, n_mean, embedding, False)
+        n_mean_vec = np.round(gbs.mean_photons_by_mode(params), 5)
```

Comment:
Does the test fail without `round`?

Reply:
Yes, it is used for checking that the vector is constant. But good spot, I don't like it and have changed to
`assert np.allclose(n_mean_vec - n_mean_vec[0], np.zeros(dim))`